MirrorProvider/MirrorProvider.Linux/LinuxFileSystemVirtualizer.cs

```diff
@@ -0,0 +1,247 @@
+                        {
+                        Console.WriteLine($"WriteFileContents failed: {result}");
+                            return false;
+                        }
+
+                        return true;
+                    });
+
+                if (hydrateFileResult != FileSystemResult.Success)
+                {
+                    return Result.EIOError;
+                }
+            }
+            catch (IOException e)
+            {
+                Console.WriteLine($"IOException in OnGetFileStream: {e.Message}");
+                return Result.EIOError;
+            }
+
+            return Result.Success;
+        }
+
+        private void OnFileModified(string relativePath)
+        {
+            Console.WriteLine($"OnFileModified: {relativePath}");
+        }
+
+        private Result OnPreDelete(string relativePath, bool isDirectory)
+        {
+            Console.WriteLine($"OnPreDelete (isDirectory: {isDirectory}): {relativePath}");
+            return Result.Success;
+        }
+
+        private void OnNewFileCreated(string relativePath, bool isDirectory)
+        {
+            Console.WriteLine($"OnNewFileCreated (isDirectory: {isDirectory}): {relativePath}");
+        }
+
+        private void OnFileRenamed(string relativeDestinationPath, bool isDirectory)
+        {
+            Console.WriteLine($"OnFileRenamed (isDirectory: {isDirectory}) destination: {relativeDestinationPath}");
+        }
+
+        private void OnHardLinkCreated(string relativeNewLinkPath)
+        {
+            Console.WriteLine($"OnHardLinkCreated: {relativeNewLinkPath}");
+        }
+
+        private bool TryGetSymLinkTarget(string relativePath, out string symLinkTarget)
+        {
+            symLinkTarget = null;
+            string fullPathInMirror = this.GetFullPathInMirror(relativePath);
+
+            const ulong BufSize = 4096;
+            byte[] targetBuffer = new byte[BufSize];
+            long bytesRead = ReadLink(fullPathInMirror, targetBuffer, BufSize);
+            if (bytesRead < 0)
+            {
+                Console.WriteLine($"GetSymLinkTarget failed: {Marshal.GetLastWin32Error()}");
+                return false;
+            }
+
+            targetBuffer[bytesRead] = 0;
+            symLinkTarget = Encoding.UTF8.GetString(targetBuffer);
+
+            if (symLinkTarget.StartsWith(this.Enlistment.MirrorRoot, StringComparison.OrdinalIgnoreCase))
+            {
+                // Link target is an absolute path inside the MirrorRoot.
+                // The target needs to be adjusted to point inside the src root
+                symLinkTarget = Path.Combine(
+                    this.Enlistment.SrcRoot.TrimEnd(Path.DirectorySeparatorChar),
+                    symLinkTarget.Substring(this.Enlistment.MirrorRoot.Length).TrimStart(Path.DirectorySeparatorChar));
+            }
+
+            return true;
+        }
+
+        private static byte[] ToVersionIdByteArray(byte version)
```

Comment:
Same comment for these two methods, does it make sense to move them to a common spot? @sanoursa, and preferences here?

Reply:
My own preference is to keep the MirrorProvider simple. Rather than refactoring for maximum code reuse, this simpler layout makes it easier to understand how any given platform works.

Creating a common helper class could be fine. Refactoring with base classes and other complications like we're planning to do in VFS for Git seems overkill for the MirrorProvider.

Reply:
That sounds good to me thanks

Reply:
Thanks, @wilbaker and @sanoursa ... and I should note again that I expect I'll have adjustments as we actually start to wire up methods.